File: tests/scripts/requestsTests/deleteTests.cpp
#include "utils.cpp"

void	deleteTest(void)
{
	printHeader("\t File Case");
	testServers("Normal Case", "/delete/full/classic", "\t\t\t\t\t");
	testServers("Normal Case (\"../\" in URL)","/delete/full/../../delete/full/noback", "\t");
	testServers("No Perm file","/delete/full/noperms", "\t\t\t\t\t");
	testServers("No Perm directory of file","/delete/cant/tryme", "\t\t\t\t");
	testServers("Only Read Perm directory of file","/delete/readme/deleteme", "\t\t");
	testServers("Not Found","/delete/emptwswdy", "\t\t\t\t\t\t");

	printHeader("\t Directory Case");
	testServers("Url not end by \"/\"","/delete/folder/empty", "\t\t\t\t");
	testServers("Empty Directory","/delete/folder/empty/", "\t\t\t\t");
	testServers("Normal Case","/delete/folder/classic/", "\t\t\t\t\t");
	testServers("No Perms for parent dir","/delete/folder/nopermspa/", "\t\t\t");
	testServers("No Perms for the dir do del","/delete/folder/noperms/", "\t\t\t");
	testServers("Dir in Dir (But have no perms)","/delete/folder/dire/", "\t\t\t");
	testServers("Dir in Dir (Have no perms but empty)","/delete/folder/dire2/", "\t\t");
	testServers("Dir in Dir (Read Only but empty)","/delete/folder/dire3/", "\t\t");
	testServers("Normal Case ++++++","/delete/folder/dire4/", "\t\t\t\t");
}

int	main(void)
{
	if (std::system("curl -I localhost:8181 > /dev/null 2>&1") != 0)
	{
		std::cout << "Nginx server not running, exiting" << std::endl;
		return (EXIT_FAILURE);
	}
	if (std::system("curl -I localhost:8080 > /dev/null 2>&1") != 0)
	{
		std::cout << "Webserv server not running, exiting" << std::endl;
		return (EXIT_FAILURE);
	}

	if (std::system("mkdir -p ../../../code/unitTest && cd ../../../code/unitTest && ../../tests/scripts/requestsTests/initDeleteTest.sh") != 0  // For our server
		|| std::system("cd ../../website && ../scripts/requestsTests/initDeleteTest.sh") != 0)  // For nginx
	{
		std::cout << "Error executing the delete init scripts" << std::endl;
		return (EXIT_FAILURE);
	}

	deleteTest();


	if (std::system("cd ../../../code/unitTest && ../../tests/scripts/requestsTests/cleanDeleteTest.sh") != 0 // For our server
		|| std::system("cd ../../website && ../scripts/requestsTests/cleanDeleteTest.sh") != 0) // For nginx
	{
		std::cout << "Error executing the delete cleaning scripts" << std::endl;
		return (EXIT_FAILURE);
	}
	return (EXIT_SUCCESS);
}

File: tests/scripts/requestsTests/deleteTests.mjs
import { compareGoodRequests } from "./compareRequests.mjs"
import { verifyServersAreRunning, exec, printHeader, COLOR_GREEN, COLOR_RED, COLOR_RESET } from "./utils.mjs"

let	failedTests = [];

async function	runGoodDeleteTest(target)
{
	const result = await compareGoodRequests(target, "DELETE", null, {});
	if (result == true)
		console.log(COLOR_GREEN + "[OK] " + COLOR_RESET);
	else
	{
		console.log(COLOR_RED + "[KO] " + COLOR_RESET);
		failedTests.push(target);
	}
	return (result);
}

async function runTests()
{
	printHeader("Normal Case");
	await runGoodDeleteTest("/delete/full/classic");

	printHeader("Normal Case (\"../\" in URL)");
	await runGoodDeleteTest("/delete/full/../../delete/full/noback");

	printHeader("No Perm file");
	await runGoodDeleteTest("/delete/full/noperms");

	printHeader("No Perm directory of file");
	await runGoodDeleteTest("/delete/cant/tryme");

	printHeader("Only Read Perm directory of file");
	await runGoodDeleteTest("/delete/readme/deleteme");

	printHeader("Not Found");
	await runGoodDeleteTest("/delete/emptwswdy");

	printHeader("Url not end by \"/\"");
	await runGoodDeleteTest("/delete/folder/empty");

	printHeader("Empty Directory");
	await runGoodDeleteTest("/delete/folder/empty/");

	printHeader("Normal Case");
	await runGoodDeleteTest("/delete/folder/classic/");

	printHeader("No Perms for parent dir");
	await runGoodDeleteTest("/delete/folder/nopermspa/");

	printHeader("No Perms for the dir do del");
	await runGoodDeleteTest("/delete/folder/noperms/");

	printHeader("Dir in Dir (But have no perms)");
	await runGoodDeleteTest("/delete/folder/dire/");

	printHeader("Dir in Dir (Have no perms but empty)");
	await runGoodDeleteTest("/delete/folder/dire2/");

	printHeader("Dir in Dir (Read Only but empty)");
	await runGoodDeleteTest("/delete/folder/dire3/");

	printHeader("Normal Case ++++++");
	await runGoodDeleteTest("/delete/folder/dire4/");


	if (failedTests.length == 0)
		printHeader("Everything Done : " + COLOR_GREEN + "[OK] " + COLOR_RESET);
	else
	{
		printHeader("Everything Done : " + COLOR_RED + "[KO] " + COLOR_RESET);
		console.table(failedTests);
	}
}

async function	run()
{
	verifyServersAreRunning();

	exec("cd ../../webserv && ../scripts/requestsTests/initDeleteTest.sh"); // setup webserv
	exec("cd ../../nginx && ../scripts/requestsTests/initDeleteTest.sh"); // setup nginx

	await runTests();

	exec("cd ../../webserv && ../scripts/requestsTests/cleanDeleteTest.sh"); // clean webserv
	exec("cd ../../nginx && ../scripts/requestsTests/cleanDeleteTest.sh"); // clean nginx
}

run();
